### PR TITLE
[Doc] Update metaprogramming doc

### PR DESCRIPTION
**Pull request by erizmr**
Related issue = #

### CODE REVIEW

File: docs/lang/articles/advanced/meta.md [hunk 1]
```diff
@@ -4,92 +4,91 @@ sidebar_position: 1
...
-- Simplify the development of Taichi standard library
+- Enabling the development of dimensionality-independent codes, e.g., codes which are
+  adaptive for both 2D/3D physical simulations.
+- Improving runtime performance by moving computations from runtime to compile time.
+- Simplifying the development of Taichi standard library.
 
-Taichi kernels are _lazily instantiated_ and a lot of computation can
-happen at _compile-time_. Every kernel in Taichi is a template kernel,
-even if it has no template arguments.
+:::note Taichi kernels are **lazily instantiated** and massive computations can
```
Comment (victoriacity) [inline docs/lang/articles/advanced/meta.md:14]:
```suggestion
:::note Taichi kernels are **lazily instantiated** and large amount of computation can
```

File: docs/lang/articles/advanced/meta.md [hunk 2]
```diff
@@ -4,92 +4,91 @@ sidebar_position: 1
...
+be executed at **compile-time**. Every kernel in Taichi is a template kernel,
+even if it has no template arguments. :::
 
 ## Template metaprogramming
 
-You may use `ti.template()` as a type hint to pass a field as an
-argument. For example:
+By using `ti.template()` as a argument type hint, a Taichi field can be passed into a kernel. Template programming also enable the codes to be reused for fields with different shape:
```
Comment (victoriacity) [inline docs/lang/articles/advanced/meta.md:20]:
```suggestion
By using `ti.template()` as a argument type hint, a Taichi field can be passed into a kernel. Template programming also enables the codes to be reused for fields with different shapes:
```

File: docs/lang/articles/advanced/meta.md [hunk 3]
```diff
@@ -4,92 +4,91 @@ sidebar_position: 1
...
 
 ```python {2}
 @ti.kernel
-def copy(x: ti.template(), y: ti.template()):
+def copy1D(x: ti.template(), y: ti.template()):
```
Comment (victoriacity) [inline docs/lang/articles/advanced/meta.md:24]: Shall we use snake case such as `copy_1D` or `copy_1d`?

File: docs/lang/articles/advanced/meta.md [hunk 4]
```diff
@@ -4,92 +4,91 @@ sidebar_position: 1
...
 
-As shown in the example above, template programming may enable us to
-reuse our code and provide more flexibility.
 
 ## Dimensionality-independent programming using grouped indices
 
-However, the `copy` template shown above is not perfect. For example, it
-can only be used to copy 1D fields. What if we want to copy 2D fields?
-Do we have to write another kernel?
+Taichi provides `ti.grouped` syntax which supports grouping loop indices into a `ti.Vector`. It enables dimensionality-independent programming, i.e., codes are adaptive to scenarios of different dimensions automatically:
```
Comment (victoriacity) [inline docs/lang/articles/advanced/meta.md:43]:
```suggestion
Taichi provides `ti.grouped` syntax which supports grouping loop indices into a `ti.Vector`.
It enables dimensionality-independent programming, i.e., codes are adaptive to scenarios of
different dimensionalities automatically:
```

File: docs/lang/articles/advanced/meta.md [hunk 5]
```diff
@@ -4,92 +4,91 @@ sidebar_position: 1
...
-:tada: Not necessary! Taichi provides `ti.grouped` syntax which enables you to
-pack loop indices into a grouped vector to unify kernels of different
-dimensionalities. For example:
+@ti.kernel
+def copy3d(x: ti.template(), y: ti.template()):
+    for i, j, k in x:
+        y[i, j, k] = x[i, j, k]
 
-```python {3-10,15-16}
+# Kernels listed above can be unified by one kernel using `ti.grouped`:
```
Comment (victoriacity) [inline docs/lang/articles/advanced/meta.md:61]:
```suggestion
# Kernels listed above can be unified into one kernel using `ti.grouped`:
```

File: docs/lang/articles/advanced/meta.md [hunk 6]
```diff
@@ -98,34 +97,27 @@ For sparse fields, the full domain shape will be returned.
...
 
-Getting the number of matrix columns and rows will allow you to write
-dimensionality-independent code. For example, this can be used to unify
-2D and 3D physical simulators.
-
-`matrix.m` equals to the number of columns of a matrix, while `matrix.n`
-equals to the number of rows of a matrix. Since vectors are considered
-as matrices with one column, `vector.n` is simply the dimensionality of
-the vector.
+For matrices, `matrix.m` and `matrix.n` returns the number of columns and rows respectively.
```
Comment (victoriacity) [inline docs/lang/articles/advanced/meta.md:100]:
```suggestion
For matrices, `matrix.m` and `matrix.n` returns the number of columns and rows, respectively.
```

File: docs/lang/articles/advanced/meta.md [hunk 7]
```diff
@@ -98,34 +97,27 @@ For sparse fields, the full domain shape will be returned.
...
+For vectors, they are treated as matrices with one column in Taichi, `vector.n` is the number of elements of the vector.
```
Comment (victoriacity) [inline docs/lang/articles/advanced/meta.md:101]:
```suggestion
For matrices, `matrix.m` and `matrix.n` returns the number of columns and rows, respectively.
For vectors, they are treated as matrices with one column in Taichi, where `vector.n` is the number of elements of the vector.
```

File: docs/lang/articles/advanced/meta.md [hunk 8]
```diff
@@ -98,34 +97,27 @@ For sparse fields, the full domain shape will be returned.
...
 
 ## Compile-time evaluations
 
-Using compile-time evaluation will allow certain computations to happen
-when kernels are being instantiated. This saves the overhead of those
-computations at runtime.
+Using compile-time evaluation will allow certain computations to be executed when kernels are being instantiating. This helps the compiler to conduct optimization and reduce the
+certain computations overhead at runtime:
 
 - Use `ti.static` for compile-time branching (for those who come from
```
Comment (victoriacity) [inline docs/lang/articles/advanced/meta.md:119]:
```suggestion
- Use `ti.static` for compile-time branching (for those who are familiar with
```

Comment (victoriacity) [inline docs/lang/articles/advanced/meta.md:119]:
```suggestion
Using compile-time evaluation allows for some computation to be executed when kernels are instantiated. This helps the compiler to conduct optimization and reduce
computational overhead at runtime:

- Use `ti.static` for compile-time branching (for those who are familiar with
```

File: docs/lang/articles/advanced/meta.md (hunk 2 above)
Comment (erizmr) [reply docs/lang/articles/advanced/meta.md:20]: Good catch!

File: docs/lang/articles/advanced/meta.md (hunk 3 above)
Comment (erizmr) [reply docs/lang/articles/advanced/meta.md]: Have modified the function names to the snake case.

**Review by victoriacity — APPROVED** [review]
LGTM!

**Pull request merged**